rubix-core/src/main/java/com/qubole/rubix/core/CachingFileSystem.java

```diff
@@ -58,26 +60,89 @@
 public abstract class CachingFileSystem<T extends FileSystem> extends FilterFileSystem
 {
   private static final Log log = LogFactory.getLog(CachingFileSystem.class);
+  private static BookKeeperFactory bookKeeperFactory = new BookKeeperFactory();
+
+  // statics that need to initialized after the configuration object is available
+  private static final AtomicBoolean initialized = new AtomicBoolean(false);
```

Comment:
I think we can use a `volatile` reference for `ClusterManager` or `CachingFileSystemStatsProvider` and do null check on those to avoid the need for `AtomicBoolean initialized`

Reply:
volatile will not help with synchronisation, we need to take a lock on some non-null static object to initialise other static members

Reply:
did you mean `volatile AtomicReference`? That could be done, I confused it with normal variable declaration.

Reply:
I meant that synchronization can be done on null check of volatile reference to object instead of a new AtomicBoolean

Reply:
There are more than one static variables getting initialised. Using a common flag `initialised` is helping avoid multiple checks, one for  each static variables, to declare initialisation complete.
Using just one of the static variables to declare initialisation complete is error prone as one variable initialised successfully doesn't always guarantee others are too, we can check last static but that too can get into trouble if order of initialisation changes in future.